src/basics.ml

```diff
@@ -239,3 +251,120 @@ let term_of_rhs : rule -> term = fun r ->
+  let compare : t -> t -> int = fun a b ->
+    let rec loop (acc:int) = function
+      | []      , []                    -> acc
+      | h1 :: t1, h2 :: t2 when h1 = h2 -> loop 0 (t1, t2)
+      | h1 :: _ , h2 :: _               -> Pervasives.compare h1 h2
+      | _ :: _  , []                    -> 1
+      | []      , _ :: _                -> ~-1 in
+    loop 0 (a, b)
+
+  (** [pp o p] output position [p] to channel [o]. *)
+  let pp : t pp = fun oc pos ->
+    match pos with
+    | [] -> Format.fprintf oc "ε"
+    | x  -> List.pp (fun oc -> Format.fprintf oc "%d") "." oc (List.rev x)
+
+  (** Initial position. *)
+  let init = []
+
+  (** [succ p] returns the successor of position [p].  For instance, if
+      [p = [1 ; 1]], [succ p = [2 ; 1]].  The succ of the initial position is
+      the first subterm of this position. *)
+  let succ = function
+    | []      -> 0 :: init
+    | x :: xs -> succ x :: xs
+
+  (** [prefix p q] sets position [p] as prefix of position [q], for instance,
+      {i prefix 1 3.4} is {i 1.3.4}; which is represented [prefix [1] [4;3]]
+      is [[4;3;1]]. *)
+  let prefix : t -> t -> t = fun p q -> q @ p
+
+  (** [sub p] returns the position of the first subterm of [p]. *)
+  let sub : t -> t = fun p -> 0 :: p
+
+  (** [tag ?empty a] attaches the positions to an array of terms [a] as if
+      they were the subterms of a same term.  If [?empty] is supplied, the
+      first element of the list is at position [empty]. *)
+  let tag : ?empty:t -> term array -> (term * t) array = fun ?empty xs ->
+    let start, p = match empty with
+      | None | Some([]) -> 0, init
+      | Some(s :: p)    -> s, p in
+    Array.mapi (fun i e -> (e, prefix p [i + start])) xs
+end
+
+(** Functional map with [Subterm.t] as keys *)
+module SubtMap = Map.Make(Subterm)
+
+(** {3 Operators on trees} *)
+
+(** [iter l n f t] is a generic iterator on trees; with function [l] performed
+    on leaves, function [n] performed on nodes, [f] returned in case of
+    {!constructor:Fail} on tree [t]. *)
+let tree_iter :
+  do_leaf:(int IntMap.t -> (term_env, term) Bindlib.mbinder -> 'a) ->
+  do_node:(int -> bool -> 'a TcMap.t -> 'a option -> 'a) ->
+  do_fetch:(bool -> 'a -> 'a) ->
+  fail:'a -> tree -> 'a = fun ~do_leaf ~do_node ~do_fetch ~fail t ->
+  let rec loop = function
+    | Leaf(pa, a)                                 -> do_leaf pa a
+    | Fail                                        -> fail
+    | Node({ swap ; store ; children ; default }) ->
+       do_node swap store
+         (TcMap.map (fun c -> loop c) children)
+         (Option.map loop default)
+    | Fetch(store, next)                          -> do_fetch store (loop next)
+  in
+  loop t
+
+(** [capacity t] computes the capacity of tree [t].  During evaluation, some
+    terms that are being filtered by the patterns need to be saved in order to
+    be bound in the right hand side of the rule.  The capacity is an upper
+    bound of the number of terms to be saved. *)
```

Comment:
This code seems very complicated. Could you please mathematically define what is the capacity of a tree?